lib/client.js

```diff
@@ -31,115 +55,146 @@ module.exports = function (dependencies) {
      }, this.config.heartBeat).unref();
   }
 
-  Client.prototype.write = function write (notification, device, count) {
-    // Connect session
-    if (!this.session || this.session.destroyed) {
-      this.session = http2.connect(`https://${this.config.address}`, this.config);
+  Client.prototype._createSession = function() {
+    const url = this._mockOverrideUrl || `https://${this.config.address}`;
+    // Get the reference to the current session so that
+    // we don't unintentionally destroy a different session on an async callback
+    const session = http2.connect(url, this.config);
 
-      this.session.on("socketError", (error) => {
-        if (logger.enabled) {
-          logger(`Socket error: ${error}`);
-        }
-        if (this.session && !this.session.destroyed) {
-          this.session.destroy();
-        }
+    session.on("socketError", (error) => {
+      if (logger.enabled) {
+        logger(`Socket error: ${error}`);
+      }
+      safeCloseSession(session);
+    });
+    session.on("error", (error) => {
+      if (logger.enabled) {
+        logger(`Session error: ${error}`);
+      }
+      safeCloseSession(session);
+    });
+
+    session.on("goaway", (errorCode, lastStreamId, opaqueData) => {
+      logger(`GOAWAY received: (errorCode ${errorCode}, lastStreamId: ${lastStreamId}, opaqueData: ${opaqueData})`);
+      // gracefully stop accepting new streams
+      // This may be redundant, since nodejs http2 client is supposed to shut down automatically on receiving a goaway frame
+      safeCloseSession(session);
```

Comment:
same here

Reply:
Done, found a few more places where this.session should have only be cleaned up if it was the session that encountered the issue.